pkg/kubelet/kubelet.go

```diff
@@ -821,6 +812,21 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
 	klet.softAdmitHandlers.AddPodAdmitHandler(lifecycle.NewNoNewPrivsAdmitHandler(klet.containerRuntime))
 	klet.softAdmitHandlers.AddPodAdmitHandler(lifecycle.NewProcMountAdmitHandler(klet.containerRuntime))
 
+	// Any admission handler that depends on GetNode() (or getNodeAway()) must have its addition delayed
+	// after the node sync with the apiserver. This is done to prevent caching of Node objects.
+	go func() {
+		klog.InfoS("Waiting for node sync before adding a predicate admission handler")
+		for {
+			if nodeHasSynced() {
+				break
+			}
+			time.Sleep(config.WaitForAPIServerSyncPeriod)
+		}
+		klog.InfoS("Adding predicate admission handler")
+		criticalPodAdmissionHandler := preemption.NewCriticalPodAdmissionHandler(klet.GetActivePods, killPodNow(klet.podWorkers, kubeDeps.Recorder), kubeDeps.Recorder)
+		klet.admitHandlers.AddPodAdmitHandler(lifecycle.NewPredicateAdmitHandler(klet.getNodeAnyWay, criticalPodAdmissionHandler, klet.containerManager.UpdatePluginResources))
```

Comment:
hmm, hadn't noticed this async addition before.

`AddPodAdmitHandler` doesn't look threadsafe, will it case data race panics if this is appended while pod admission is being used to admit static pods?

Also, how does adding this late affect admission of static pods?

Reply:
> AddPodAdmitHandler doesn't look threadsafe, will it case data race panics if this is appended while pod admission is being used to admit static pods?

well, i'm not seeing a panic in my tests at least.

@ehashman had a similar question here.
https://github.com/kubernetes/kubernetes/pull/99336#discussion_r585889220

i would assume a critical section around this append would suffice:
https://github.com/kubernetes/kubernetes/blob/cea1d4e20b4a7886d8ff65f34c6d4f95efcb4742/pkg/kubelet/lifecycle/interfaces.go#L105

> Also, how does adding this late affect admission of static pods?

i think that is a question for @derekwaynecarr

Reply:
@neolit123 @liggitt static pods do not go through kubelet admission checks.

see:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L2095

one can argue if this is a good thing or not, but after thinking through this PR more, i think we can do the following:

remove this special casing admission logic now that we know we are holding sync of apiserver sourced pods until the node sync has occurred elsewhere.